rfcs/function-and-table-types.md

```diff
@@ -0,0 +1,68 @@
+# Add primitive function and table types
+
+## Summary
+
+Add types for "real" functions and tables.
+
+## Motivation
+
+Some APIs require "real" functions and tables, not just things that
+"look functiony" (e.g. tables with a `__call` metamethod) or "look
+tably" (e.g. instances of classes). This RFC adds types for those.
+
+For example, the function:
+
+```lua
+  function succ(x)
+    if type(x) == "function"
+      return x() + 1
+    else
+      assert(type(x) == number)
+      return x + 1
+    end
+  end
+```
+
+cannot quite be given an accurate Luau type. The nearest is `(number |
+()->number) -> number` but this is slightly too generous, since it
+allows "functiony" types such as tables with a `__call` metamethod:
+
+```lua
+  local t = setmetatable({}, {__call = function(self) return 5 end})
+  succ(t)
+```
+
+This will typecheck but produce a runtime error.
+
+A similar issue affects tables.
+
+The common cases in practice are built-in APIs such as the Luau
+standard library, or APIs implemented using the C++ FFI. For example,
+`pairs` takes a table argument, and gives a runtime error if it is called
+with a class instance.
+
+## Design
+
+Add:
+
+* a type `table`, inhabited by Luau tables (but not class instances), and
+* a type `function`, inhabited by Luau functions (but not class methods or
+tables with metamethods).
+
+Luau functions with known source and target types are now an intersection type `function & (T) -> U`.
```

Comment:
With my remark above about callable tables not being subtypes of current function types, can we go in other direction and define `(T) -> U` to be a real function type like it works right now, provide `(T) -> U | Callable<(T) -> U>` as an opt in for places where you want to accept callable tables and leave `function` just as the supertype of function types as was suggested by some developers?

Reply:
Maybe even just `Callable<(T) -> U>` instead of `(T) -> U | Callable<(T) -> U>`.

Reply:
or a trait/type family/whatever we use for overloaded operators for `Callable`